### PR TITLE
Enable testing with MySQL and Postgres on Travis-CI

### PR DESCRIPTION
**Pull request by atodorov**
This PR enables MySQL and Postgres on Travis. I have a support ticket opened with their team to figure out how to add MariaDB alongside MySQL.

The build will fail b/c there are some issue with both MySQL and Postgres. I will try to fix them in subsequent commits.

### CODE REVIEW

**Comment by atodorov** [comment]
@tkdchen sorry for the noise here. I've managed to fix the failing tests on MySQL but I'm having troubles configuring MariaDB to use utf8 instead of latin1. I have opened a support ticket for that.

**Comment by tkdchen** [comment]
It's okay. Thanks for making this improvement. Looking forward to the success.

**Comment by atodorov** [comment]
@tkdchen this is ready for merge. 

I have update the Travis CI config to use their Ubuntu Trusty infrastructure b/c I was having lots of problems with MariaDB on their older infrastructure.

I have also updated some of the tests, mostly to avoid duplicate objects so they now PASS.

Note1: I have cherry-picked #157 here and later modified it to special case the migration for PostgreSQL. This turned out to be a bug in Django 1.8, which is documented and already fixed (reference in comments and commit log) but not backported. Also note that we can't use RunSQL for SQLite here b/c the necessary syntax isn't supported. Django ORM goes behind the scenes and provides the migration for us (add column, copy data, drop column, rename, etc).

Note 2: Currently tests for PostgreSQL fail. At least one of them is related to my other PRs where I refactor the SQL. My vote is to merge this PR as-is and I will work to add more tests to my other PRs and work to fix the PostgreSQL failures there.

**Comment by atodorov** [comment]
@tkdchen ping? Any comment on this PR. I can't proceed further without feedback.

**Comment by tkdchen** [comment]
@atodorov Hi, I'm just back from vocation. I'll be back to review PRs later.

**Comment by atodorov** [comment]
Depends on https://github.com/Nitrate/Nitrate/pull/157

**Comment by atodorov** [comment]
@tkdchen I think this is ready to merge so we can continue forward. I hate to merge with failing Postgres but MySQL, MariaDB and sqlite all work fine and some of the fixes are covered into other PRs. I don't think we have a lot of other options here otherwise this PR will become very big and there are other problems to worry about too.

**Comment by tkdchen** [comment]
This is a big step forward to the testing. However, the problem is there is dependency of Travis in Nitrate source code, and this testing matrix cannot run locally by developer. I prefer to a solution to avoid these two major disadvantages.

Basically, the testing matrix should

- allow developer to run tests locally without any other external services required.
- allow developer to run tests by selecting specific database, SQLite, MySQL, MariaDB or PostgreSQL, or even all of them at once whatever in parallel or in sequence.
- by default, tests could be run with SQLite, which is a much fast way to run tests and verify something quickly.

The second ability should be able to be used in Travis-CI to run the testing matrix.

In my mind, docker would be a good tool to help us to build this testing matrix, and obviously some scripts would be required.

Currently, to run whole test suite, I can simple run ``make test`` in a well-provisioned vagrant machine. However, running test suite from scratch should be an easy thing now in Nitrate as following steps,

1. provision machine, installing necessary packages required in next steps
1. make a virtual environment
1. install dependencies from ``requirements/devel.txt``
1. ``make test``

It should be easy to make these steps happen in a docker container.  Fedora 25 docker image can be used and it is ok to use the version of MySQL, MariaDB and PostgreSQL in Fedora repo.

Thank you for making this improvement. It's valuable and develop our ideas. Let's make it better and better.

I'll look at other commits in this PR from tomorrow. Have a good day. :)

**Comment by atodorov** [comment]
>
    make test, run with SQLite by default
    DB=[dbname] make test, dbname could be mysql, mariadb, or pgsql
    DB=all make test, run tests with mysql, mariadb, and pgsql individually

I think this can easily be done in the settings and then simply ask Travis to execute whichever command fits its matrix. In this way we move the infrastructure dependency in `.travis.yml` and not in the project source code. 

I will rework the settings for that. It is up to the developer to prepare their local environment for whichever DB they want to test on. I will also update the docs with commands how to install dependencies and run the tests against different databases.

Note: vagrant or docker will be a great help but that's up to each developer to decide how to use. IMO this needs to come at a later stage because this PR is quite big already. What do you think?

**Comment by atodorov** [comment]
@tkdchen see the updated documentation and the last 3 commits for the latest changes. It is not possible to execute the tests on the local machine by specifying which database to use. The same commands are integrated into Travis as well. All of the integration happens in .travis.yml.

**Comment by tkdchen** [comment]
Using Travis to run tests does not mean there is a dependency of Travis to run tests. Simply say I don't want this line of code appears in source code

```
if 'TRAVIS' in os.environ:
```

Documentation is good but not always. For running tests, I prefer to make it as easy as possible with just one command, no need of steps

1. find the documentation
1. read through the documentation to learn the steps
1. try each step to provision environment and launch tests
1. repeat again and again if something is wrong by misoperation

Instead, as mentioned in my previous comment, just ``make test`` or ``DB=pgsql make test`` is enough. Easy and save much time, isn't it?

If someone is interested in the steps to build testing environment, just open and read the relative scripts. That should be more straightforward than lots of words in documentation and easy understand.

**Comment by atodorov** [comment]
> if 'TRAVIS' in os.environ:

Mistake, this has now been updated. The rest (make test commands) are as requested and described in more details in the docs. Anything else missing ?

**Comment by tkdchen** [comment]
@atodorov We talked a lot in IRC channel #nitrate. You did a good job to make it possible to run tests on different database engines. Hopefully, you could continue improving based on the changes in this PR to make running tests more easily.

I'd like to repeat my thought again

1. Do not depend on Travis-CI to run tests.
1. To achieve the first goal, if those database provision steps in ``.travis.yml`` are necessary, they should happen locally, not in Travis-CI.
1. As a developer, I can run tests simply in following steps **without the need of reading whole documentation before that**.
    1. clone Nitrate
    1. ``DB=mysql make test``

Detailed documentation in this PR is good, I think it could be as reference for developers who are interested in hacking the tests for themselves.

Is this clear and make sense?

**Comment by tkdchen** [comment]
I merged these 3 commits

a1c47ec Compare PKs with what was written in DB, not hard-coded
fc2343c Update pre_check_product() and pre_process_ids() to include long
7d10b49 Don't use unique test object names

File: tcms/testruns/tests.py [hunk 1]
```diff
@@ -37,7 +36,7 @@ def setUpTestData(cls):
         cls.client = test.Client()
 
     def test_404_if_run_does_not_exist(self):
-        nonexisting_run_pk = TestRun.objects.count() + 1
+        nonexisting_run_pk = 999999
```
Comment (tkdchen) [inline tcms/testruns/tests.py]: It would be helpful to understand if you can add explanation to this change. Then, I'll merge it.

Comment (atodorov) [inline tcms/testruns/tests.py]: OK, for some reason ( I think because PKs are global autoincrements on MySQL, generated via LAST_INSERT_ID()) 
```
TestRun.objects.count() == 1
nonexisting_run_pk == 2
```

but the only TestRun object in the database has an actual ID of 2 hence the failure!

On other DB engines the only TestRun object has an ID of 1.  I will update this code with
`TestRun.objects.last().pk + 1` which will be more robust.

However I have found one more thing. In `testruns/urls/run_urls.py` there are two URLs which use the same view:
`.../ordercase/$` and `.../ordercaserun/$`. Do you know why is that? Sounds like a bug (let me know and I will send a separate PR for it).

**Comment by atodorov** [comment]
Updated and rebased.

@tkdchen let's move this comment https://github.com/Nitrate/Nitrate/pull/171#issuecomment-294270654 into a separate issue. I understand the value proposition of it and it is definitely doable even in the short term. However not everyone would like to follow exactly the same steps so we shouldn't force it. Plus a separate issue will allow us to discuss into more details. What do you think?

**Comment by tkdchen** [comment]
This was done in 2e22b8a09b3 and subsequent commits that make it possible to run with PostgreSQL.
